Agent spec process (#304)
elastic/apm: docs/agents/spec-process.md

Every change that affects more than one agent should be initiated via a change to the specification rather than creating a cross-agent issue.

> ## Advantages

Reviewer: nit: I'd put the `Advantages` part at the end. I feel long term everyone will be interested in the process itself and not about the advantage of it.

- Ensures we have an up-to-date [spec](https://www.joelonsoftware.com/2000/08/09/the-joel-test-12-steps-to-better-code/) for agents.
- Easier to keep track of what the current state of the proposal is.
- Comments to a specific sentence of the spec can be made inline,
  just like in code reviews.
- If there are changes or amendments to a spec later on,
  a PR to an existing spec is much cleaner than having to track down all preceding issues to find out what the current state is.
- Easier to get a new agent (in-house or community) started.

## Process

1. Agents discussion issue \

>   If there's a lot of uncertainty about what the poposal is,

Reviewer: ```suggestion
  If there's a lot of uncertainty about what the proposal is,
```

  a discussion issue should be created.
  In doubt, lean towards not creating a discussion issue and start with a draft PR.
  Discussion issues should not have votes/agent checkboxes.
  It's not expected that all agents participate in the initial discussions, although everyone is invited to do so.
1. Draft PR \
  It's fine if the first draft of the spec change even if details are unclear and need discussion.
  The PR should include which agents are affected by this change.
  Specifically, it should always be clear whether the RUM agent is affected by this.
  This doesn't have to be determined yet when creating the draft PR,
  but it is required before marking the PR as `Ready for Review`.
1. Approval by at least one other agent \
  This makes sure the change makes sense for not just one agent.

>   This doesn't ping all agent teams yet,

Reviewer: The only unclear part to me about this step is: how do I get someone to approve the draft? The draft PR can be approved, but if we don't ping the team, then is the person who open the PR expected to just somehow find someone from another agent team to review the PR (e.g. on slack)?

Author: I think the spec creator should actively reach out to another team. Ideally, they already know another team that has gotten a similar request. Otherwise, they can ask for a design partner on slack or on a weekly. If all of that fails, a lead has to step in.

  which only happens in the next step.
  Of course,
  every agent is more than welcome to contribute at any point in this process,
  but they are not expected to actively keep track if it.
1. Ready for Review \
  This automatically requests review from all agents due to the [`CODEOWNERS`](https://github.com/elastic/apm/tree/master/.github/CODEOWNERS) file.

>   After a quorum of agents has approved, others have 1w to veto by requesting changes.

Reviewer 2: ```suggestion
  After a quorum of agents has approved, others have 1week to veto by requesting changes.
```

was confusing on first read.

>   After a quorum of agents has approved, others have 1 week to veto by requesting changes.

Reviewer: To this part:
> After a quorum of agents has approved

This sound very nice in theory, but I'm not totally sure it'll wok well in practice. I personally would not even be able to define what quorum of agents is off the top of my head. Why don't we just say "After at least 1 other agent approved" (or any other reasonable number). The draft needs to be approved by 1 other agent, this would be another one and probably the person who opens the PR also does it for a specific agent, so just by 1 additional approve here we'd already cover 3 agents at this point.

I know it's extremely simple, but I feel that's easier to follow and won't make a huge difference for the outcome.

Author: Good idea. Before changing I'd like to hear a few more opinions. @roncohen also raised some concerns about this part.

Author: Should we maybe require one more agent and a PM to approve? Then we'd have at least a total of 3 agent's and 1 PM's approval.

This makes sure PMs are always in the loop. Often times PMs would request a feature. If they don't see the exact proposal there's a chance the proposal doesn't do what they had in mind.
In other cases, a dev might have proposed a spec proposal based on a request form the community. In this case, it's also important to let PMs know about those changes.
Obviously, additional devs and leads can always approve or object at any given time.

The balance I'm trying to strike is to make sure everyone is in the loop while minimizing noise and not making the process too heavyweight.

Author: Seems to be the only outstanding question. I've changed it to what I've described in the previous comment. Happy to revisit though.

A counter-argument would be that this may introduce a bottleneck on the PM side. However, we already need their input for prioritization anyway.

Reviewer: I think this is fine. This will force us a bit to have more discussions with PMs which is a good thing; especially in this case, because these are real world changes in agents, not only high level discussions. If this turns out to be a bottleneck we can still change it later.

>   If the proposed changes are likely to be problematic for a certain agent,

Author: It's mostly fine if a change doesn't apply to, or makes sense for, a particular agent.

  an approval from that agent is mandatory.
1. Prioritization \
  At any point in time,
  but before merging the PR,
  the priority should be determined.
  That happens in collaboration with devs and leads.
  It's recommended to do that sooner rather than later
  so that we don't spend a lot of time discussing a feature that's not going to be prioritized.
  As a result of the prioritization,
  the PR gets assigned to a milestone,
  reflecting the stack release version this feature should be implemented by all agents.
  If some agents are not going to implement the change by that stack release version,
  or if there are no immediate plans to implement the change at all,
  a comment should be made on the PR.
1. PR can be merged by the author of the PR \
  Three business days before merging, there should be a last call for objections.

> 1. The author of the PR creates an issue in each agent's repo \

Author: Does that create too much burden on the initiator of a change and may lead to fewer change requests? Should the team lead be responsible for merging and creating the implementation issues?

Author: Yep, that would be great!
But I'd decouple that so we start out with the manual approach and then, possibly with help from our friendly robots, automate that step.
This process isn't set in stone after merging it. Just like code, we can always change the process by creating a PR 🙂

Reviewer 2: Gotcha, totally 👍

Reviewer: ➕ on the script.

> Does that create too much burden on the initiator of a change and may lead to fewer change requests? Should the team lead be responsible for merging and creating the implementation issues?

I think the initiator should do this by default, I don't think that opening the issues is that big of a deal; and we can still also put a friendly message there that every agent team could open an issue and link it. If we say that the team lead does it I think that would make the process slower - let everyone do it and then ultimately script it.

Reviewer: Oh, on this note: We already have lots of agents, how do we make sure we open issues for all? I feel it's very easy to accidentally miss some agents. Can't we maybe still have a list with checkbox here? Maybe in the PR description.

Author: Good point. We can add a table for each agent's issue but I wouldn't add checkboxes as I'd prefer voting via approving the PR.

Author: Added this to the last step

  They make sure the issues are assigned to a stack release milestone,
  based on the milestone of the spec PR.
  Agent teams can still change the milestone if priorities shift.

>   However, they should talk to their team lead about it.

Reviewer: ```suggestion
  However, this should be discussed in the specific agent's planning meeting and mentioned in the agent (bi-)weekly meeting.
```

I think instead of chasing a team lead, it's better to have a discussion in the agent meeting, we sometimes also have PMs there and if we mention it in the agents weekly, then the person who opened the PR will be also informed about it. Of course team lead can jump in at any time.

Author: The issue I see is that currently, the agent meetings are bi-weekly. So there could be 2 weeks before discussing an issue might slip into another release. Also, it's likely not too relevant for agent team A that agent team B implements a change a bit later.

Reviewer: I totally agree on the bi-weekly part and the low relevance to other teams (that's the 2. part of the suggestion). My main point (the 1. part of the suggested change) was that I prefer these to be team decisions and not be by default a single person decision because in my experience that introduces bottleneck and does not improve the quality of the decision. 

Also the specific agent's planning meeting is in my opinion a really good setup for these decisions, we have products + dev + maybe a team lead there. I feel a team lead can always jump in at any point and force a decision (they don't need special permissions in the process, they have it anyways :) ); so rather than leaving this by default to a team lead, I'd try to encourage the specific agent team with all other stakeholders (eg. product) to make a reasonable decision in the agent's planning meeting.

Author: > I prefer these to be team decisions and not be by default a single person decision because in my experience that introduces bottleneck and does not improve the quality of the decision.

Totally agree. It should mainly be the team's decision. They should just let their team lead know which could be in the form of a slack message or in a weekly meeting. I wouldn't want to be too specific to allow for flexibility.

What do you think of changing the wording to 
> However, they should let their team lead know.

The thought behind it is that the team lead can then inform PMs if necessary so that devs don't have to chase them. I wouldn't want to assume that there's always a PM on the agent weeklies.

Reviewer: > What do you think of changing the wording to 
>> However, they should let their team lead know.

That sounds very good to me. 👍

  If the proposed changes are likely to be problematic for a certain agent,
  an approval from that agent is mandatory.
1. Prioritization \
  At any point in time,
  but before merging the PR,
  the priority should be determined.
  That happens in collaboration with devs and leads.
  It's recommended to do that sooner rather than later
  so that we don't spend a lot of time discussing a feature that's not going to be prioritized.
  As a result of the prioritization,
  the PR gets assigned to a milestone,
  reflecting the stack release version this feature should be implemented by all agents.
  If some agents are not going to implement the change by that stack release version,
  or if there are no immediate plans to implement the change at all,
  a comment should be made on the PR.
1. PR can be merged by the author of the PR \
  Three business days before merging, there should be a last call for objections.
1. The author of the PR creates an issue in each agent's repo \
  They make sure the issues are assigned to a stack release milestone,
  based on the milestone of the spec PR.

>   The author of the PR adds a comment to the PR with a table that lists each agent and the link to the implementation issue.

Reviewer 3: Can we provide a link to a sample code block for copy-paste here? Tables are a pain in markdown.

Author: great idea, added

> Every change that affects more than one agent should be initiated via a change to the specification rather than creating a cross-agent issue.

Reviewer 4: ```suggestion
Every change that affects more than one agent should be initiated via a change to the specification.  Once accepted, cross-agent issues can follow.
```

Reviewer 4: I updated the suggestion, I meant per-agent, not cross-agent

Author: As this is described in more detail in the last step, I wouldn't want to anticipate it here.
If you find the term `a cross-agent issue` ambiguous, we change it to this:

```suggestion
Every change that affects more than one agent should be initiated via a PR against the specification rather than creating an issue that describes the changes in its description.
```

## Process

1. Agents discussion issue \
  If there's a lot of uncertainty about what the proposal is,
  a discussion issue should be created.
  In doubt, lean towards not creating a discussion issue and start with a draft PR.
  Discussion issues should not have votes/agent checkboxes.

>   It's not expected that all agents participate in the initial discussions, although everyone is invited to do so.

Reviewer 4: I'd highlight sooner that a draft PR is the preferred first step.  Something like:
```
Open a draft PR to change the specification to initiate a discussion.  If discussion is required before a spec change proposal can even be assembled, create an Agent discussion issue first ...
```

Author: done. better now?

1. Draft PR \

>   It's fine if the first draft of the spec change even if details are unclear and need discussion.

Reviewer 4: is there an example of a good spec change PR to link to here?

  The PR should include which agents are affected by this change.
  Specifically, it should always be clear whether the RUM agent is affected by this.
  This doesn't have to be determined yet when creating the draft PR,

>   but it is required before marking the PR as `Ready for Review`.

Reviewer 4: any other suggestions for what makes a good spec change?  Some ideas:
* minimum version of the elastic stack required for this, in case of new intake. 
* link to related meta issues

Author: I think that part is TBD. It's certainly an important part of the spec but I wonder if we could define that in a follow-up. I think after going through a couple of iterations of this process we can give better guidance.

> minimum version of the elastic stack required for this, in case of new intake.

Wouldn't that always be the version the spec PR is assigned to? If not, would you add that to the PR description or in the agent spec itself?

Reviewer 4: Tackling this in a follow up sounds fine as does capturing version information in the milestone.  We always have the option to add that later if things are unclear.

  The author of the PR adds a comment to the PR with a table that lists each agent and the link to the implementation issue.
  This makes it easier to see the current implementation state for each agent and makes it easy to verify that there's an issue for each agent.
  Agent teams can still change the milestone if priorities shift.
  However, they should let their team lead know.

## Advantages

- Ensures we have an up-to-date [spec](https://www.joelonsoftware.com/2000/08/09/the-joel-test-12-steps-to-better-code/) for agents.
- Easier to keep track of what the current state of the proposal is.
- Comments to a specific sentence of the spec can be made inline,
  just like in code reviews.
- If there are changes or amendments to a spec later on,
  a PR to an existing spec is much cleaner than having to track down all preceding issues to find out what the current state is.
- Easier to get a new agent (in-house or community) started.

## Guiding principles

- Everyone in the team should be empowered to propose changes
- The process should be lightweight and painless 
- Keep the noise low by notifying the right people at the right time and maturity stage of the proposal

> If it turns out that it's not accomplishing that it's a bug,

Reviewer 4: I think a word or two are missing here?

Author: better now?

  As a result of the prioritization,
  the PR gets assigned to a milestone,
  reflecting the stack release version this feature should be implemented by all agents.
  If some agents are not going to implement the change by that stack release version,
  or if there are no immediate plans to implement the change at all,
  a comment should be made on the PR.
1. PR can be merged by the author of the PR \
  Three business days before merging, there should be a last call for objections.
1. The author of the PR creates an issue in each agent's repo \
  They make sure the issues are assigned to a stack release milestone,
  based on the milestone of the spec PR.
  The author of the PR adds a comment to the PR with a table that lists each agent and the link to the implementation issue.
  A template for this table can be found in the next section.
  This makes it easier to see the current implementation state for each agent and makes it easy to verify that there's an issue for each agent.
  Agent teams can still change the milestone if priorities shift.
  However, they should let their team lead know.

## Agent issue table template

> | Agent   | Link to agent implementation issue |
> | --------|------------------------------------|
> | .NET    | 
> | Go      | 
> | Java    | 
> | Node.js | 
> | PHP     | 
> | Python  | 
> | Ruby    | 
> | RUM     | 

Reviewer 5: ```suggestion
\```
| Agent   | Link to agent implementation issue |
| --------|------------------------------------|
| .NET    | 
| Go      | 
| Java    | 
| Node.js | 
| PHP     | 
| Python  | 
| Ruby    | 
| RUM     | 
\```
```

Failing to properly format this in MD but my suggestions was to keep the copy-pastable table in a code block for even easier copy-paste-ability.

## Advantages

- Ensures we have an up-to-date [spec](https://www.joelonsoftware.com/2000/08/09/the-joel-test-12-steps-to-better-code/) for agents.
- Easier to keep track of what the current state of the proposal is.
- Comments to a specific sentence of the spec can be made inline,
  just like in code reviews.
- If there are changes or amendments to a spec later on,
  a PR to an existing spec is much cleaner than having to track down all preceding issues to find out what the current state is.
- Easier to get a new agent (in-house or community) started.

## Guiding principles

- Everyone in the team should be empowered to propose changes
- The process should be lightweight and painless 
- Keep the noise low by notifying the right people at the right time and maturity stage of the proposal

> If it turns out that this process is not accomplishing these goals that it's a bug,

Reviewer 6: ```suggestion
If it turns out that this process is not accomplishing these goals then it's a bug,
```
